hyperspy/signal.py

```diff
@@ -3288,6 +3314,8 @@ def diff(self, axis, order=1, out=None, rechunk=True):
         >>> s.diff(-1).data.shape
         (64,64,1023)
         """
+        if not self.axes_manager[axis].is_linear:
+            raise NonLinearAccessError()
```

Comment:
Similar to the `sum` versus `integrate_simpson` point mentioned above, I would say that `diff` does not require a linear axis, but the current implementation of the `derivative` requires linear axis.
It may be worth clarifying this in the docstring (and user guide?).

Reply:
The original implementation required a linear axis, as the axis was rewritten based on the offset. I have now simply added that for a non-linear axis the last point of the axis is dropped instead.
By the way, integrate_simpson works correctly for a non-linear axis as well (as long as during the conversion the intensities have been rescaled correctly).
I added a Note to the docstring for every function that has an exception raised. In the user guide we should mention a few specific functions, and `derivative` is a good example.